driver-core/src/main/com/mongodb/internal/logging/StructuredLogger.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.logging;
+
+import com.mongodb.connection.ClusterId;
+import com.mongodb.internal.VisibleForTesting;
+import com.mongodb.internal.diagnostics.logging.Logger;
+import com.mongodb.internal.diagnostics.logging.Loggers;
+import com.mongodb.internal.logging.StructuredLogMessage.Entry;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.mongodb.internal.VisibleForTesting.AccessModifier.PRIVATE;
+import static java.lang.String.format;
+
+/**
+ * <p>This class is not part of the public API and may be removed or changed at any time</p>
+ */
+public class StructuredLogger {
+
+    private final Logger logger;
+
+    private static final Map<String, StructuredLoggingInterceptor> INTERCEPTORS = new HashMap<>();
+
+    public static void addInterceptor(final String applicationName, final StructuredLoggingInterceptor interceptor) {
```

Comment:
This is here so that the tests can intercept the log messages and make assertions.

Reply:
I have used `@VisibleForTesting(otherwise = PRIVATE)` in such cases previously, and I think that's perfectly fine. An example: `Timeout.startNanos()`.

Reply:
At first glance, it may apprear that one could introduce a "virtual" `AccessModifier.NONE` to convey that a program element would not even exist if it were not for tests, but I don't think this adds any useful information: if we have a `@VisibleForTesting(otherwise = PRIVATE)` element and observe that it is used neither in tests nor in the declaring class (I am obviously ignoring outer classes here, but the idea is the same), then we know it can be deleted. What could `AccessModifier.NONE` inform us about on top of that? It could only "prevent" us from using the annotated program element within the declaring class. However, such prevention does not make sense: the code in a class can do whatever it wants, the `@VisibleForTesting` is not there to restrict that; this annotation is supposed restrict the usage of exposed program elements *outside* of the class.